### PR TITLE
FileHandle: _readDataOfLength: reduce fstat() calls

### PR DESCRIPTION
**Pull request by spevans**
- Cache the blocksize and a regular file flag to avoid
  repeated calls to fstat().

### CODE REVIEW

**Comment by spevans** [comment]
@swift-ci test

File: Foundation/FileHandle.swift [hunk 1]
```diff
@@ -211,7 +211,46 @@ open class FileHandle : NSObject {
...
+        }
+        return statbuf
+    }
+
+    private func _fileSize() throws -> Int64 {
+        let statbuf = try _fstat()
+        return Int64(statbuf.st_size)
+    }
+
+    private var _isRegularFile: Bool? = nil
```
Comment (compnerd) [inline Foundation/FileHandle.swift]: Can this not change?  If I call the method and underneath remove the file and recreate it, the cached value will be wrong.  I don't think it is sufficient to just cache this value.

File: Foundation/FileHandle.swift [hunk 2]
```diff
@@ -211,7 +211,46 @@ open class FileHandle : NSObject {
...
+    private var _blockSize: Int? = nil
```
Comment (compnerd) [inline Foundation/FileHandle.swift]: This is unlikely to change unless the volume is changed.  Do we need to consider any side effects from operations like `chroot` altering the volume while the application is executing?

File: Foundation/FileHandle.swift (hunk 1 above)
Comment (spevans) [reply Foundation/FileHandle.swift]: If the file descriptor is already open, deleting/recreating the underlying file wont change what fd is pointing at.

Although I just noticed there is a `closeFile` method which should invalidate the cached values so I will update it for that.

**Comment by millenomi** [comment]
In general, I am skeptical of optimizations without a target, especially when the source of truth is the filesystem. What is this reduction useful for?

Also: you have a time-of-check-time-fo-use race window between the `_isFile…` call on 333 and the `mmap` immediately below, which given how volatile `mmap` can be is concerning.

**Comment by spevans** [comment]
The file size is not cached and always read just before the mmap(). Im pretty sure once an FD has been opened its filetype and block size don't change unless Im mistaken?

**Comment by shahmishal** [comment]
The Swift project moved the default branch to `main` and deleted `master` branch, so GitHub automatically closed the PR. Please re-create pull request with main branch.

More detail about the branch update - https://forums.swift.org/t/updating-branch-names/40412

**Pull request closed without merging**